synapse/lib/remcycle.py

```diff
@@ -0,0 +1,944 @@
+          "definition" which must contain the ingest definition. The ingest
+          definition must contain a "open" directive which is used with the
+          ingest iterdata() function to process the API data prior to ingest.
+        * vars: This is a dictionary of items which are stamped into the url
+          template during the construction of the Nyx object using format().
+
+    See a complete example below:
+
+    ::
+
+        {
+          "api_args": [
+            "someplace"
+          ],
+          "api_optargs": {
+            "domore": 0
+          },
+          "doc": "api example",
+          "http": {
+            "headers": {
+              "token-goodness": "sekrit token"
+            }
+          },
+          "ingest": {
+            "definition": {
+              "ingest": {
+                "forms": [
+                  [
+                    "inet:ipv4",
+                    {
+                      "var": "ip"
+                    }
+                  ]
+                ],
+                "vars": [
+                  [
+                    "ip",
+                    {
+                      "path": "ip"
+                    }
+                  ]
+                ]
+              },
+              "open": {
+                "format": "json"
+              }
+            },
+            "name": "geolocv4"
+          },
+          "url": "http://vertex.link/api/v4/geoloc/{{someplace}}/info?domore={{domore}}&apikey={APIKEY}",
+          "vars": {
+            "APIKEY": "8675309"
+          }
+        }
+
+    ::
+
+    This example should be interpreted as the following:
+
+        1. The APIKEY value in the 'vars' will be set in the URL, resulting in
+           the following default url:
+
+        ::
+
+            "http://vertex.link/api/v4/geoloc/{someplace}/info?domore={domore}&apikey=8675309"
+
+        ::
+
+        2. The HTTP request will have the header "token-goodness" set to
+           "sekrit token" for the request.
+        3. The caller must provide the "someplace" value in the api_args when
+           calling buildHttpRequest.  The caller may provide the "domore"
+           value if they want to override the default value of "0".
+        4. An Ingest will be created for parsing the data from the API and
+           made available to the Hypnos object.
+
+    :param config: API Endpoint configuration outlined above.
+    '''
+
+    def __init__(self, config):
+        self._raw_config = dict(config)
+        self.required_keys = ['url',
+                              'doc'
+                              ]
+        self.doc = ''
+        self.url_template = ''
+        self.url_vars = {}
+        self.effective_url = ''
+        self.request_defaults = {}
+        self.api_args = []
+        self.api_kwargs = {}
+        self.gest = None
+        self.gest_name = None
+        self.gest_open = None
+
+        self._parseConfig()
+
+    @property
```

Comment:
@property functions haven't traditionally worked well with things like telepath.  ( it gets confused by you trying to access a "local" rather than call something ).  We avoid them in synapse typically.